include/openthread/tcp.h

```diff
@@ -0,0 +1,639 @@
+ * deallocates all resources it has internally allocated for this TCP endpoint.
+ * The application can reuse the memory backing the TCP endpoint as it sees fit.
+ *
+ * If it corresponds to a live TCP connection, the connection is terminated
+ * unceremoniously (as in otTcpAbort()). All resources the application has
+ * provided for this TCP endpoint (linked buffers for the send buffer, memory
+ * for the receive buffer, the @p aEndpoint structure itself, etc.) are
+ * immediately returned to the application.
+ *
+ * @param[in]  aEndpoint  A pointer to the TCP endpoint structure to deinitialize.
+ *
+ * @retval OT_ERROR_NONE    Successfully deinitialized the TCP endpoint.
+ * @retval OT_ERROR_FAILED  Failed to deinitialize the TCP endpoint.
+ */
+otError otTcpEndpointDeinitialize(otTcpEndpoint *aEndpoint);
+
+struct otTcpListener;
+typedef struct otTcpListener otTcpListener;
+
+typedef enum otTcpIncomingConnectionAction {
+    OT_TCP_INCOMING_CONNECTION_ACTION_ACCEPT,
+    OT_TCP_INCOMING_CONNECTION_ACTION_DEFER,
+    OT_TCP_INCOMING_CONNECTION_ACTION_REFUSE,
+} otTcpIncomingConnectionAction;
+
+/**
+ * This callback indicates that an incoming connection that matches this TCP
+ * listener has arrived.
+ *
+ * The typical response is for the application to accept the incoming
+ * connection. It does so by populating @p aAcceptInto with a pointer to the
+ * otTcpEndpoint into which to accept the incoming connection. This
+ * otTcpEndpoint must already be initialized using otTcpEndpointInitialize().
+ * Then, the application returns OT_TCP_INCOMING_CONNECTION_ACTION_ACCEPT.
+ *
+ * Alternatively, the application can decline to accept the incoming
+ * connection. There are two ways for the application to do this. First, if the
+ * application returns OT_TCP_INCOMING_CONNECTION_ACTION_DEFER, then OpenThread
+ * silently ignores the connection establishment request; the connection peer
+ * will likely retransmit the request, at which point the callback will be
+ * called again. This is valuable if resources are not presently available to
+ * accept the connection, but they may be available when the connection peer
+ * retransmits its connection establishment attempt. Second, if the application
+ * returns OT_TCP_INCOMING_CONNECTION_ACTION_REFUSE, then OpenThread sends a
+ * "connection refused" message to the host that attempted to establish a
+ * connection. If the application declines the incoming connection, it is not
+ * required to populate @p aAcceptInto.
+ *
+ * @param[in]   aListener    The TCP listener that matches the incoming connection.
+ * @param[in]   aPeer        The host and port from which the incoming connection originates.
+ * @param[out]  aAcceptInto  The TCP endpoint into which to accept the incoming connection.
+ *
+ * @retval  Description of how to handle the incoming connection.
+ */
+typedef otTcpEndpoint *(*otTcpAcceptReady)(otTcpListener *aListener, const otSockAddr *aPeer, otTcpEndpoint **aAcceptInto);
```

Comment:
I think return type should be `otTcpIncomingConnectionAction`?

Reply:
Yes, thanks for catching that. I fixed this in the latest commit.